### PR TITLE
feat: Flesh out rewards with a base class & add a new reward fn: FrequentAssetRewardFn.

### PR DESCRIPTION
**Pull request by anordin95**
Changes:
- Set up RewardFn as an abstract base class rather than a TypeAlias. I felt that may clarify the process of making a custom one i.e. subclassing. 
- I also modified the reward function API to be (prior-obs, prior-action, obs). I _think_, but am not sure, that may be the more common approach for stochastic environments and/or multi-agent scenarios. I also lay out two good reasons for doing so in the doc-string of the RewardFn base class.
- Implement the WinLoss reward
- Implement another reward: FrequentAssetRewardFn. It provides frequent rewards (i.e. most turns will see a non-zero reward) based on the change in assets, i.e. land, army, cities.

Let me know what ya think!

### CODE REVIEW

File: generals/rewards/common.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from generals.core.observation import Observation
+
+
+def compute_num_cities_owned(observation: Observation) -> int:
+    owned_cities_mask = observation.cities & observation.owned_cells
+    num_cities_owned = owned_cities_mask.sum()
+    return num_cities_owned
+
+
+def compute_num_generals_owned(observation: Observation) -> int:
```
Comment (strakam) [inline generals/rewards/common.py:16]: I now recall that you once asked whether this repo supports FFA and I forgot to reply. Currently it does not, and I don't plan to do it personally, unless some external force makes me :D So in this repository this is just 1 general = still playing, 2 generals = win, 0 generals = lose. I think you already know it, just making a comment :)

File: generals/rewards/frequent_asset_reward_fn.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+from generals.core.action import Action
+from generals.core.observation import Observation
+from generals.rewards.common import compute_num_cities_owned, compute_num_generals_owned, is_action_valid
+from generals.rewards.reward_fn import RewardFn
+
+
+class FrequentAssetRewardFn(RewardFn):
```
Comment (strakam) [inline generals/rewards/frequent_asset_reward_fn.py:7]: this is very nice, i was thinking about something similar myself. It is a good start on how people should think about reward shaping for this game :)

File: generals/rewards/frequent_asset_reward_fn.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        change_in_num_generals_owned = compute_num_generals_owned(obs) - compute_num_generals_owned(prior_obs)
+        # Moderately reward valid actions & penalize invalid actions.
+        valid_action_reward = 1 if is_action_valid(prior_action, prior_obs) else -5
+
+        reward = (
+            valid_action_reward
+            + 1 * change_in_army_size
+            + 5 * change_in_land_owned
+            + 10 * change_in_num_cities_owned
+            + 10_000 * change_in_num_generals_owned
```
Comment (strakam) [inline generals/rewards/frequent_asset_reward_fn.py:25]: don't have an intuition on what `x10_000` does to gradients tho :D

File: generals/rewards/reward_fn.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+from generals.core.action import Action
+from generals.core.observation import Observation
+
+
+class RewardFn(abc.ABC):
+    @abc.abstractmethod
+    def __call__(self, prior_obs: Observation, prior_action: Action, obs: Observation) -> float:
+        """
+        It's common to see notation like r(s, a) in RL theory and it's accordingly common for environments
```
Comment (strakam) [inline generals/rewards/reward_fn.py:11]: thats true (that it is very common), but i feel like in practice the post-action observation is even more important and i even see it in literature, so I think this is much better 👍

File: generals/rewards/win_lose_reward_fn.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+from generals.core.observation import Observation
+from generals.rewards.common import compute_num_generals_owned
+from generals.rewards.reward_fn import RewardFn
+
+
+class WinLoseRewardFn(RewardFn):
+    """A simple reward function. +1 if the agent wins. -1 if they lose."""
+
+    def __call__(self, prior_obs: Observation, prior_action: Action, obs: Observation) -> float:
+        change_in_num_generals_owned = compute_num_generals_owned(obs) - compute_num_generals_owned(prior_obs)
```
Comment (strakam) [inline generals/rewards/win_lose_reward_fn.py:11]: stylish :D

**Review by strakam — APPROVED** [review]
Hi Anordin!

this is honestly a great PR :) It provides a much easier way to create custom reward functions (up until now, users had to bend many things to achieve it i think).

I'm currently working on a supervised agent and will soon start to fine-tune with RL. I'm sure I would have to do changes like this myself, so I'm happy that you were quicker and I will get to use it :D

Cheers!
Matej

**Comment by strakam** [comment]
im thinking whether it makes more sense to move the `rewards/` folder into `core/`? what do you think?

File: generals/rewards/common.py (hunk 1 above)
Comment (anordin95) [reply generals/rewards/common.py:16]: Mhm, thanks for the heads up! If I can ever get my 1v1 bot working, I may try sorting out an FFA-capable one.

File: generals/rewards/frequent_asset_reward_fn.py (hunk 2 above)
Comment (anordin95) [reply generals/rewards/frequent_asset_reward_fn.py:7]: thanks!

File: generals/rewards/frequent_asset_reward_fn.py (hunk 3 above)
Comment (anordin95) [reply generals/rewards/frequent_asset_reward_fn.py:25]: Hahaha yeah. For a vanilla gradient descent technique that may be problematic. But, for something like Adam which normalizes with an exponentially decaying running average of past updates I _think_ it might be ok? 

I wanted to choose a number high enough that even after a few hundred turns (i.e. $$\gamma^{300}$$) it'd dominate all other reward components. Right now my gamma/time-discount-factor is 0.995 so 10,000 may be a bit overkill!

File: generals/rewards/reward_fn.py (hunk 4 above)
Comment (anordin95) [reply generals/rewards/reward_fn.py:11]: Mmm cool! I've been reading the chapter on RL from https://udlbook.github.io/udlbook/. I know folks highly recommend Sutton & Barto -- is that what you're reading/read?

Comment (strakam) [reply generals/rewards/reward_fn.py:11]: Yeah, I've been reading that and it is a very cool book. I get most of my knowledge from our uni course tho. It is epic :D https://ufal.mff.cuni.cz/courses/npfl139/2324-summer

**Comment by strakam** [comment]
Let's keep it as it is then (the `rewards/` location). Regarding supervised agent, I'll dm u on discord :)

File: generals/rewards/reward_fn.py (hunk 4 above)
Comment (anordin95) [reply generals/rewards/reward_fn.py:11]: Mmm nice. Looks good! I noticed you share a last name with the instructor -- are y'all related?
